src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php

```diff
@@ -485,8 +505,81 @@ protected function castAttribute($key, $value)
                 return $this->asDateTime($value);
             case 'timestamp':
                 return $this->asTimestamp($value);
-            default:
-                return $value;
+        }
+
+        return $value;
+    }
+
+    /**
+     * Cast the given attribute to a class.
+     *
+     * @param  string  $key
+     * @return mixed
+     */
+    protected function castToClass($key)
+    {
+        if (isset($this->classCastCache[$key])) {
+            return $this->classCastCache[$key];
+        }
+
+        list($type, $attributes) = $this->getClassCast($key);
+
+        if ($attributes) {
+            $parameters = array_map(function ($key) {
+                return $this->attributes[$key] ?? null;
+            }, $attributes);
+        } else {
+            $parameters = [$this->attributes[$key] ?? null];
+        }
+
+        return $this->classCastCache[$key] = $this->{'castTo'.Str::studly($type)}(...$parameters);
+    }
+
+    /**
+     * Determine whether a value is class castable.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isClassCastable($key)
+    {
+        $casts = array_map(function ($cast) {
+            return strpos($cast, ':') === false ? $cast : strstr($cast, ':', true);
+        }, $this->getCasts());
+
+        if (! array_key_exists($key, $casts)) {
+            return false;
+        }
+
+        return method_exists($this, 'castTo'.Str::studly($casts[$key]));
+    }
+
+    /**
+     * Merge the cast class attributes back into the model.
+     *
+     * @return void
+     */
+    protected function mergeAttributesFromClassCasts()
+    {
+        foreach ($this->getCasts() as $attribute => $cast) {
+            if ($this->isClassCastable($attribute)) {
+                list($type, $attributes) = $this->getClassCast($attribute);
+
+                $attributeCount = count($attributes);
+                $object = $this->castToClass($attribute);
+
+                $castedAttributes = method_exists($this, 'castFrom'.Str::studly($type))
+                    ? $this->{'castFrom'.Str::studly($type)}($object)
+                    : $object->__toString();
+
+                if ($attributeCount !== count($castedAttributes)) {
+                    throw new LogicException("Class cast {$attribute} must return {$attributeCount} attributes");
```

Comment:
Wouldn't be better to change this to:
"Class cast {$attribute} must return {$attributeCount} attribute(s)" ? I've seen it in some parts of the core